### PR TITLE
Remove mustcopy and add AbstractSolverParameter

### PR DESCRIPTION
**Pull request by blegat**
See discussion following https://github.com/JuliaOpt/MathOptInterfaceTests.jl/pull/44#issuecomment-359189967

### CODE REVIEW

**Comment by blegat** [comment]
Here, all variables and constraints attribute are obviously removed with the variables. Does it mean that all variable and constraint attributes should have `mustcopy` equal to `true` ?
Then, should we move `AbstractInstanceAttribute` that have `mustcopy` equal to `false` simply be `AbstractSolverAttribute` and get rid of `mustcopy` ?

**Comment by mlubin** [comment]
> Then, should we move AbstractInstanceAttribute that have mustcopy equal to false simply be AbstractSolverAttribute and get rid of mustcopy ?

@blegat I'm having trouble coming up with a reason not to do this. It matches up with Gurobi's concept of a parameter, so we might as well call it `AbstractSolverParameter`. Should the MOIU instance support setting arbitrary parameters?

@joaquimg @odow

**Comment by joaquimg** [comment]
We have to be careful if one set a solver1 parameter and then calls solver2. If parameters are integers MOIU might not break and silently set a completely different parameter. Solvers should have ther AbstractGurobiSolverParameter or we make these concrete GurobiSolverParameter(parameter,value)...

**Comment by blegat** [comment]
Each parameter will have its own type, subtype of `AbstractSolverParameter`, when parameters will be passed from a solver to another solver through a `copy!`, all solver specific parameters will be dropped since the other solver will return `false` to `canset` since it does not know the parameter.

**Comment by blegat** [comment]
@mlubin @joaquimg I have redirected the PR to introduce AbstractSolverParameter and remove mustcopy. What do you think of it ?

Note that `ListOfSolverParametersSet` is a `AbstractSolverParameter` since it is not affected by `empty!`.

File: src/attributes.jl [hunk 1]
```diff
@@ -21,9 +33,13 @@ Abstract supertype for attribute objects that can be used to set or get attribut
 """
 abstract type AbstractConstraintAttribute end
 
-const AnyAttribute = Union{AbstractInstanceAttribute, AbstractVariableAttribute, AbstractConstraintAttribute}
+const AnyAttribute = Union{AbstractSolverParameter, AbstractInstanceAttribute, AbstractVariableAttribute, AbstractConstraintAttribute}
```
Comment (mlubin) [inline src/attributes.jl]: Is a parameter a special kind of attribute or is it a different thing? We should be precise about this.

Comment (blegat) [reply src/attributes.jl]: I think it is better to split but still have a common supertype

**Comment by blegat** [comment]
For default implementations of `get`, `set!`, ... I have included the parameter inside an existing method since otherwise that would unallow an instance to do e.g. `Union{AbstractSolverParameter, AbstractInstanceAttribute}` since it would give an ambiguity.

File: src/attributes.jl [hunk 2]
```diff
@@ -201,11 +229,18 @@ Assign a value to the attribute `attr` of constraint `c` in instance `instance`.
...
+    throw(ArgumentError("AbstractInstance of type $(typeof(instance)) does not support setting the $(_name(attr))"))
 end
 
-# TODO: solver-independent parameters
+## Solver parameters
+
+"""
+    ListOfSolverParametersSet()
+
+A `Vector{AbstractSolverParameters}` of all solver parameters that were set to the instance.
```
Comment (joaquimg) [inline src/attributes.jl]: Should be sigular, right?

Comment (blegat) [reply src/attributes.jl]: Good catch !

File: docs/src/apireference.md [hunk 3]
```diff
@@ -11,6 +11,7 @@ CurrentModule = MathOptInterface
 List of attribute categories.
 
 ```@docs
+AbstractSolverParameter
```
Comment (mlubin) [inline docs/src/apireference.md]: The documentation needs restructuring (parameters aren't attributes).

**Comment by blegat** [comment]
I have restructured the doc :)

**Pull request merged**